Standalone Pre-Publish Classification Panel

faisal-alvi opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

This PR fixes a regression where the "Classify Post" suggestion in the Pre-Publish panel was no longer visible after recent refactoring. The issue occurred because the `PrePubClassifyPost` component was embedded within the `ClassificationButton` component instead of being registered as a separate WordPress plugin.

**Solution:**
- Extracted the pre-publish panel into a separate plugin file (`pre-publish-panel.js`)
- Added proper webpack entry and PHP enqueue for the new component
- Removed the embedded usage from the classification button
- Reused existing code without rewriting to minimize risk

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #879

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

1. Enable and configure the Classification feature in WordPress admin
2. Create a new post and add a title and some content
3. Click on "Publish" to open the Pre-Publish panel
4. Verify the "Classify Post" suggestion appears in the Pre-Publish panel
5. Test the classification functionality by clicking "Suggest terms & tags"
6. Verify the classification button in the editor toolbar still works correctly
7. Check for any console errors in the browser developer tools

The "Classify Post" suggestion should be visible in the Pre-Publish panel, and both the pre-publish panel and classification button should function independently without conflicts.

### Changelog Entry

> Fixed - Restored "Classify Post" suggestion in Pre-Publish panel after regression

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @iamdharmesh @faisal-alvi 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added [Critical Flows, Test Cases, and/or End-to-End Tests](https://10up.github.io/Open-Source-Best-Practices/testing/) to cover my change.
- [ ] All new and existing tests pass.

faisal-alvi committed:
Standalone Pre-Publish Classification Panel

faisal-alvi committed:
js linting fixes,
/home/runner/work/classifai/classifai/src/js/features/classification/pre-publish-panel.js
Error:    23:1   error  Delete `····`                                                                                                                                                                                                prettier/prettier
Error:    26:69  error  Replace `·` with `⏎↹↹↹`                                                                                                                                                                                      prettier/prettier
Error:    29:1   error  Delete `↹↹`                                                                                                                                                                                                  prettier/prettier
Error:   178:9   error  Replace `{·__(·'Get·AI-powered·suggestions·for·categories·and·tags.',·'classifai'·)·}` with `⏎↹↹↹↹↹↹{·__(⏎↹↹↹↹↹↹↹'Get·AI-powered·suggestions·for·categories·and·tags.',⏎↹↹↹↹↹↹↹'classifai'⏎↹↹↹↹↹↹)·}⏎↹↹↹↹↹`  prettier/prettier

faisal-alvi committed:
a few more jslint fixes,
Error:    26:10  error  Replace `⏎↹↹↹getCurrentPostType,⏎↹↹↹getCurrentPostId,⏎↹↹↹getPostTypeLabel,⏎↹↹}·=·` with `·getCurrentPostType,·getCurrentPostId,·getPostTypeLabel·}·=⏎↹↹↹`  prettier/prettier
Error:   183:8   error  Replace `⏎↹↹↹↹↹↹↹` with `·`                                                                                                                                prettier/prettier
Error:   185:1   error  Delete `↹`                                                                                                                                                 prettier/prettier
Error:   186:8   error  Delete `↹`                                                                                                                                                 prettier/prettier
Error:   187:7   error  Replace `↹)⏎↹↹↹↹↹↹` with `)·`                                                                                                                              prettier/prettier

faisal-alvi committed:
final jslint fixes

faisal-alvi commented:
Ready for review.

iamdharmesh reviewed (commented):
Thanks for the PR @faisal-alvi. Code looks good. However, while testing, I noticed that the suggested terms are not being assigned to the post when it is requested from the pre-publish panel. Another issue I observed is that the "Classify Post" suggestion always appears, even when the post has already been classified and has terms saved or suggested. It should only appear if classification hasn’t been done yet.

Could you please check it once. Thanks!

faisal-alvi committed:
Fix term embedding handling to ensure it is always an array before processing

faisal-alvi committed:
Ensure embeddings are always processed as arrays

faisal-alvi committed:
Add save functionality for taxonomy terms in Pre-Publish panel

faisal-alvi committed:
eslint fix,
/home/runner/work/classifai/classifai/src/js/features/classification/pre-publish-panel.js
Error:   230:15  error  Replace `·?·__(·'Saved!',·'classifai'·)·` with `⏎↹↹↹↹↹↹?·__(·'Saved!',·'classifai'·)⏎↹↹↹↹↹↹`  prettier/prettier

faisal-alvi committed:
eslint: spacing fix

faisal-alvi commented:
@dharmesh Thanks for the feedback! I think I missed the Save button earlier. I’ve now added it, so you should be able to click Save, and the suggested terms will be correctly assigned to the post when requested from the pre-publish panel.

<img width="331" height="831" alt="image" src="https://github.com/user-attachments/assets/c8dfd553-4c4f-4c53-831f-7a28a50fa386" />

Regarding the other issue you mentioned, the “Classify Post” suggestion always appears regardless of whether the user has already added tags or categories. I believe this is intended behavior, but I think we can add some other enhancements around this area for sure. But for now, this PR focuses on bringing the ClassifAI Post tab back to the Pre-Publish panel. We can explore enhancements in future tickets.

---

@dkotter While working on this, I noticed that AI suggestions for tags and categories were intermittently failing. After debugging, I found that sometimes the $embeddings variable was populated with a single array, and other times with multiple nested arrays. I’ve added a potential fix to normalize this behavior. Although this fix isn’t originally part of the scope of this PR, it felt too minor to raise a separate PR.

The inconsistent array formation seems to stem from this part of the code:
https://github.com/10up/classifai/blob/97412809ae1ea44e71fa0855d9286de49713d92d/includes/Classifai/Features/Smart404EPIntegration.php#L214

I avoided changing this directly to prevent potential issues for existing users who may have legacy data structures in their DBs. Since it looks like you’ve worked on this area previously, I’d really appreciate your feedback on the change, especially [here](https://github.com/10up/classifai/pull/962/files/812ceb069bcaae988d3534db8bf3ba394dfb5b14..0d2564e626bf837ef36c6dc83000648ee4b54c65).

dkotter commented:
> While working on this, I noticed that AI suggestions for tags and categories were intermittently failing. After debugging, I found that sometimes the $embeddings variable was populated with a single array, and other times with multiple nested arrays. I’ve added a potential fix to normalize this behavior. Although this fix isn’t originally part of the scope of this PR, it felt too minor to raise a separate PR.
> 
> The inconsistent array formation seems to stem from this part of the code:
> 
> [classifai/includes/Classifai/Features/Smart404EPIntegration.php](https://github.com/10up/classifai/blob/97412809ae1ea44e71fa0855d9286de49713d92d/includes/Classifai/Features/Smart404EPIntegration.php#L214)
> 
> Line 214 in [9741280](https://github.com/10up/classifai/commit/97412809ae1ea44e71fa0855d9286de49713d92d)
> 
>  // Otherwise let's get all embeddings in a single request. 
> I avoided changing this directly to prevent potential issues for existing users who may have legacy data structures in their DBs. Since it looks like you’ve worked on this area previously, I’d really appreciate your feedback on the change, especially [here](https://github.com/10up/classifai/pull/962/files/812ceb069bcaae988d3534db8bf3ba394dfb5b14..0d2564e626bf837ef36c6dc83000648ee4b54c65).

So for terms, I believe we'll only ever have a single array of embeddings. But for posts, unless the post is really short, we will end up with multiple embeddings that are each stored in their own array. That should already be normalized prior to saving that data to post meta, whether we have a single array of embeddings or multiple arrays of embeddings. If that's not working properly, that is a bug we should fix when that embedding data is saved and not try and normalize that when we use the data. I haven't had time to test this myself but if you're able to reproduce this, I'd suggest we fix this during the save process and not add a normalization fix.

Depending on what is required to fix that, we can consider what to do for backwards compatibility but I'd prefer to address this in the right place if we can.

iamdharmesh commented:
@faisal-alvi I am able to see Save button now and it work as expected. So, PR is all good for me to be merged once discussion for the embeddings changes got concluded.

iamdharmesh commented:
@faisal-alvi @dkotter as pre-publish penal changes looks good, maybe we can handle the embeddings update things in separate PR to unblock the original work on this PR?

dkotter commented:
> @faisal-alvi @dkotter as pre-publish penal changes looks good, maybe we can handle the embeddings update things in separate PR to unblock the original work on this PR?

I agree, we can document this in an Issue and open a new PR to address this, and remove any code that was added to this PR to fix that

faisal-alvi committed:
Handle the embeddings update things in separate PR to unblock the original work,
https://github.com/10up/classifai/pull/962#issuecomment-3201911780
